src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+                                                      sizeof(previous.startTime), previous.policy);
+
+                index         = previousEpoch(index);
+                RootKey & old = it->second->epoch[index];
+                if (old.active)
+                {
+                    emberAfGroupKeyManagementClusterPrint("    old: key(%u), start(%u), policy:%u", sizeof(old.value),
+                                                          sizeof(old.startTime), old.policy);
+                }
+            }
+        }
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static KeySet * getKeySet(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    return it == sGroupKeys.end() ? nullptr : it->second;
+}
+
+static EmberAfStatus setKey(uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan & keyRoot, chip::ByteSpan & epochStartTime,
+                            uint8_t securityPolicy)
+{
+    if (keyRoot.size() != GROUP_KEY_ROOT_SIZE || epochStartTime.size() != GROUP_KEY_EPOCH_SIZE)
+    {
+        return EMBER_ZCL_STATUS_INVALID_VALUE;
+    }
+    else
+    {
+        KeySet * set = getKeySet(fabricIndex, keySetIndex);
+        if (set)
+        {
+            set->current = nextEpoch(set->current);
+        }
+        else
+        {
+            set                                                                  = new KeySet();
+            sGroupKeys[std::tuple<uint16_t, uint16_t>(fabricIndex, keySetIndex)] = set;
+        }
+        RootKey & key = set->epoch[set->current];
+        key.active    = true;
+        memcpy(key.value, keyRoot.data(), GROUP_KEY_ROOT_SIZE);
+        memcpy(key.startTime, epochStartTime.data(), GROUP_KEY_EPOCH_SIZE);
+        printKeys();
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+}
+
+static EmberAfStatus removeKey(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    emberAfGroupKeyManagementClusterPrint("removeKey(%u, %u)", fabricIndex, keySetIndex);
+
+    std::tuple<uint16_t, uint16_t> index(fabricIndex, keySetIndex);
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.find(index);
+    EmberAfStatus status                                            = EMBER_ZCL_STATUS_SUCCESS;
+
+    if (it == sGroupKeys.end())
+    {
+        // No key found
+        status = EMBER_ZCL_STATUS_NOT_FOUND;
+    }
+    else
+    {
+        // Existing key
+        delete it->second;
+        sGroupKeys.erase(it);
+    }
+
+    printKeys();
+    revokeKeys(fabricIndex, keySetIndex);
+    return status;
+}
+
+static EmberAfStatus removeAllKeys(uint16_t fabricIndex)
+{
+    emberAfGroupKeyManagementClusterPrint("removeAllKeys(%u)", fabricIndex);
+
+    std::map<std::tuple<uint16_t, uint16_t>, KeySet *>::iterator it = sGroupKeys.begin();
```

Comment:
`std::tuple<uint16_t, uint16_t> is repeated everywhere. Suggest using a new Struct type rather than a tuple. For a two-valued structure of basic scalar values, it's always more readable to use a structure with named fields than `tuple.first, tuple.second`